homeassistant/loader.py

```diff
@@ -1322,23 +1322,28 @@ def async_get_loaded_integration(hass: HomeAssistant, domain: str) -> Integratio
 
 async def async_get_integration(hass: HomeAssistant, domain: str) -> Integration:
     """Get integration."""
+    cache: dict[str, Integration | asyncio.Future[None]]
+    cache = hass.data[DATA_INTEGRATIONS]
+    if type(int_or_fut := cache.get(domain, _UNDEF)) is Integration:
+        return int_or_fut
     integrations_or_excs = await async_get_integrations(hass, [domain])
     int_or_exc = integrations_or_excs[domain]
-    if isinstance(int_or_exc, Integration):
+    if type(int_or_exc) is Integration:
         return int_or_exc
+    if TYPE_CHECKING:
+        assert isinstance(int_or_exc, Exception)
```

Comment:
Any particular reason to use `type` over `isinstance`? I did a performance comparison and they performed almost the same. _A few runs even had `isinstance` with a tiny lead._

`isinstance` has the advantage that it's supported by type checkers though.

```py
from timeit import timeit

setup = """\
class Integration: ...
data = Integration()
"""

print(f"isinstance: {timeit('isinstance(data, Integration)', setup=setup, number=10**8)}")
print(f"type:       {timeit('type(data) is Integration', setup=setup, number=10**8)}")
```

```
isinstance: 1.0131321661174297
type:       0.9947599170263857
```

Reply:
The type check is about the same for the hit case, but its much faster for the miss case.

```
from timeit import timeit

setup = """\
class Integration: ...
data = Integration()
"""

print(f"isinstance: {timeit('isinstance(data, Exception)', setup=setup, number=10**8)}")
print(f"type:       {timeit('type(data) is Exception', setup=setup, number=10**8)}")
```

```
isinstance: 1.6804408330353908
type:       0.8310095830238424
```

Reply:
I think in this case its very unlikely to be an Exception though so keeping it as-is won't matter